### PR TITLE
JP-4151: Give TSO3 x1dints files an S_REGION

### LINKED ISSUE #9910: No TSO L3 products in MAST JWST Search when object name specified
_Issue [JP-4151](https://jira.stsci.edu/browse/JP-4151) was created on JIRA by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly):_

For TSO products, MAST JWST Search returns level 3 data products and ASN files when program and observation is specified, but not when target name is specified. Examples attached for the following cases:
 * program 1118, obs 5, HAT-P-14
 * program 1353, obs 3, WASP-17
 * program 5863, obs 3, TOI-3884

It looks like this might be similar to the correction we made for Level 3 WFSS products: as of B12.0, we no longer generate a cal.fits product for Level 3 WFSS products, which meant no WFSS level 3 product had an associated WCS (or S_REGION). This would cause the level 3 products to have no footprint (or show up in cone searches), so we added an S_REGION to the level 3 x1d product.

We may need to investigate if a similar situation is present for TSO products.

### PR DESCRIPTION
**Pull request by emolter**
Resolves [JP-4151](https://jira.stsci.edu/browse/JP-4151)

Closes #9910 along with https://github.com/spacetelescope/stdatamodels/pull/602 

This PR adds an `s_region` attribute to the output x1dints files coming from `calwebb_tso3`.  Hopefully this allows these products to be found in MAST queries by target ID or location.

This PR handles this in the very simplest way, by just updating the `s_region` of the output to be the same as the `s_region` of the first input file.  I don't know of any tso observations where input files should have different sky footprints, but hopefully the code will fail gracefully in that case.

## Tasks
- [ ] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [x] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by emolter** [comment]
https://github.com/spacetelescope/RegressionTests/actions/runs/18472624596 pointing to stdatamodels https://github.com/spacetelescope/stdatamodels/pull/602 

All changes are expected: S_REGION header keywords are added to TSO3 x1dints products.

File: jwst/pipeline/calwebb_tso3.py [hunk 1]
```diff
@@ -192,3 +194,43 @@ def process(self, input_data):
...
+        if (len(cal_model_list) == 0) or (len(model.spec) == 0):
+            log.warning("No input or output models provided; cannot set S_REGION.")
+            return
+        has_sregion = [w.meta.wcsinfo.hasattr("s_region") for w in cal_model_list]
+        if not all(has_sregion):
+            log.warning(
+                "One or more input model(s) are missing an `s_region` attribute; "
+                "output S_REGION will not be set."
+            )
+            return
```
Comment (melanieclarke) [inline jwst/pipeline/calwebb_tso3.py]: I'm not sure this condition is possible, but if so, is it necessary to skip adding the sregion?  If we're just taking the first sregion anyway, we could just warn and use the first available?

In that case, you could also consolidate this check with the next and loop over the cal_model_list once to gather all the s_regions, instead of two separate list comprehensions.

File: jwst/pipeline/calwebb_tso3.py [hunk 2]
```diff
@@ -192,3 +194,43 @@ def process(self, input_data):
...
+
+        input_sregions = [w.meta.wcsinfo.s_region for w in cal_model_list]
+        if not all(s == input_sregions[0] for s in input_sregions):
+            log.warning(
+                "Input models have different S_REGION values; this is unexpected for tso3 data. "
+                "Setting output S_REGION to the first input model's value."
+            )
+        model.spec[0].s_region = input_sregions[0]
+        return
```
Comment (melanieclarke) [inline jwst/pipeline/calwebb_tso3.py]: Nitpick - this return is unnecessary since it's at the end of the function and no return value is expected.

File: jwst/pipeline/tests/test_calwebb_tso3.py [hunk 3]
```diff
@@ -74,3 +81,72 @@ def test_niriss_soss_full(tmp_path):
...
+
+    # no s_region attributes present
+    watcher = log_watcher(
+        "jwst.pipeline.calwebb_tso3",
+        message="One or more input model(s) are missing an `s_region` attribute;",
+        level="warning",
+    )
+    result = Tso3Pipeline()._populate_tso_spectral_sregion(model, cal_model_list)
+    watcher.assert_seen()
+    assert result is None
```
Comment (melanieclarke) [inline jwst/pipeline/tests/test_calwebb_tso3.py]: Here and in other new tests - there's no need to store or test for a result, it will always be None.  The test should just be whether `model` has an s_region populated or not.

**Review by melanieclarke — COMMENTED** [review]
Looks good, thanks!  Just a couple minor suggestions.

For the question about picking the first S_REGION - it looks to me like it might be technically possible to get an S_REGION that is not identical in the input calints files, since it is assigned from the WCS by exposure, in this function: https://github.com/spacetelescope/jwst/blob/bf7d548fff82fd8b85c7b8b001e24407034b20be/jwst/assign_wcs/util.py#L761
Combining multiple spectral targets in tso3 wouldn't make sense though, so I would think that any input S_REGION should be close enough for the purpose.  I suggest we move forward with this solution and revise later if we turn up significant exceptions.

File: jwst/pipeline/calwebb_tso3.py (hunk 1 above)
Comment (emolter) [reply jwst/pipeline/calwebb_tso3.py]: let me know if https://github.com/spacetelescope/jwst/pull/9927/commits/2dc59208ef9e12e310b7a24d22b607d486a43b70 looks better to you.  It's a bit strange-looking due to the fact that `getattr()` will set the s_region to `" "` if it doesn't exist, but hopefully not too strange-looking.

File: jwst/pipeline/calwebb_tso3.py (hunk 2 above)
Comment (emolter) [reply jwst/pipeline/calwebb_tso3.py]: should be fixed by https://github.com/spacetelescope/jwst/pull/9927/commits/2dc59208ef9e12e310b7a24d22b607d486a43b70

File: jwst/pipeline/tests/test_calwebb_tso3.py (hunk 3 above)
Comment (emolter) [reply jwst/pipeline/tests/test_calwebb_tso3.py]: should be fixed by https://github.com/spacetelescope/jwst/pull/9927/commits/2dc59208ef9e12e310b7a24d22b607d486a43b70

File: jwst/pipeline/calwebb_tso3.py (hunk 1 above)
Comment (melanieclarke) [reply jwst/pipeline/calwebb_tso3.py]: That looks fine, thanks.

**Review by melanieclarke — COMMENTED** [review]
I tested locally on a NIRSpec BOTS association with multiple segments and NRS1 and 2 detectors (jw01118-o005_20250917t232840_tso3_00001_asn.json).  For this association, all the input S_REGIONS are the same.  The output x1dints file has 6 extensions, one for each of 3 segments and 2 detectors.  The S_REGION is recorded only in the first spectrum, and matches the S_REGION from all the input files.  

I know the choice to record in the first only is consistent with the WFSS implementation, but it still seems a bit unintuitive to me -- it would be more natural to record it in either the primary HDU, or else in every EXTRACT_1D extension.  Perhaps we should file a follow up ticket to add S_REGION to all the x1d/x1dints extensions?

Either way, I think this PR is fine as is, and we can leave any further S_REGION improvements for the future.

Can you please re-run regression tests after the stdatamodels PR gets merged, just in case?

**Comment by emolter** [comment]
Can do RE regtests.

I totally agree, it's not intuitive to me either.  I don't recall exactly why the choice was made, besides that there's no SCI extension.  But given that we're not directly calling the wcsinfo subschema, I don't see any issue with just storing it in the primary.  To me, the primary header makes more sense because it's supposed to represent (at least for WFSS data) the total region over which sources were extracted.  And plus then the same value wouldn't be stored hundreds of times.

But we can discuss on the ticket.

**Comment by emolter** [comment]
https://github.com/spacetelescope/RegressionTests/actions/runs/18761792709 

these all look as expected and are ready for okify

**Comment by tapastro** [comment]
@meeseeksdev backport to release/1.20.x
